27_11_2019/SimpleImageProcessing/ppImageUtils.cpp

```diff
@@ -165,4 +250,220 @@ namespace implementation
+} // namespace implementation
+
+
+
+void test()
+{
+	Matrix dummy(2, 3, 1);
+	increase(dummy);
+	for (int i = 1; i < 3; ++i)
+		for (int j = 1; j < 4; ++j)
+		{
+			assert(dummy(i, j) == 1);
+		}
+
+	for (int m = 0; m < 5; ++m)
+	{
+		assert(dummy(0, m) == 0.0);
+	}
+	for (int i = 0; i < 5; ++i)
+	{
+		assert(dummy(3, i) == 0.0);
+	}
+	for (int i = 0; i < 4; ++i)
+	{
+		assert(dummy(i, 0) == 0.0);
+	}
+	for (int i = 0; i < 4; ++i)
+	{
+		assert(dummy(i, 4) == 0.0);
+	}
+
+	Matrix ones(3, 3, 1);
+	Matrix inp(3, 3, 0);
+	inp(0, 2) = 8;
+	inp(1, 0) = 4;
+	inp(2, 0) = 2;
+	inp(1, 1) = -2;
+	inp(2, 1) = 3;
+	inp(1, 2) = 6;
+	inp(2, 2) = 1;
+	ones(0, 1) = 0;
+	ones(1, 0) = -1;
+
+	double a = CalculateConvolution(inp, ones, 1, 1);
+	assert(a == 7.0);
+
+
+	Matrix on(3, 3, 1);
+
+	increase(inp);
+	Matrix result = masked_image_generator(inp, on);
+	assert(result(0, 0) == 2);
+	assert(result(0, 1) == 16);
+	assert(result(0, 2) == 12);
+	assert(result(1, 0) == 7);
+	assert(result(1, 1) == 22);
+	assert(result(1, 2) == 16);
+	assert(result(2, 0) == 7);
+	assert(result(2, 1) == 14);
+	assert(result(2, 2) == 8);
+
+	Matrix result_1 = row_wise_parallel_masked_image_generator(inp, on, 2);
+	//Matrix result_1 = column_wise_parallel_masked_image_generator(inp, on, 2);
+	//Matrix result_1 = cross_wise_parallel_masked_image_generator(inp, on, 2);
+	assert(result_1(0, 0) == 2);
+	assert(result_1(0, 1) == 16);
+	assert(result_1(0, 2) == 12);
+	assert(result_1(1, 0) == 7);
+	assert(result_1(1, 1) == 22);
+	assert(result_1(1, 2) == 16);
+	assert(result_1(2, 0) == 7);
+	assert(result_1(2, 1) == 14);
+	assert(result_1(2, 2) == 8);
+
+}
+
+
+
+Matrix row_wise_parallel_masked_image_generator(Matrix & input, Matrix & convolution, int number_of_threads)
+{
+
+	if (number_of_threads > input.numberOfRows() - 3)
+	{
+		throw std::runtime_error("!!!The number of threads shouldn't exceed the number of rows");
+	}
+
+
+	int size_1 = input.numberOfRows() - 2;
+	int size_2 = input.numberOfCols() - 2;
+	int numberOfThreads = number_of_threads;
+	int chunk = size_1 / numberOfThreads;
+	std::vector<std::thread> threads;
+	Matrix A = input;
+	Matrix B = convolution;
+	Matrix result(input.numberOfRows() - 2, input.numberOfCols() - 2);
+	for (int i = 0; i < numberOfThreads - 1; ++i)
+	{
+		// At each iteration, we create a new thread who gets a function to compute and store the thread in threads vector.
+		threads.emplace_back(threadwise_masking, std::ref(A), std::ref(B), chunk*i, chunk * (i + 1), size_2, std::ref(result));
+	}
+
+	// Main thread does the last chunk until the end of the arrays.
+	threadwise_masking(A, B, (numberOfThreads - 1)*chunk, size_1, size_2, result);
+
+	for (auto& thread : threads) thread.join();
+
+	// After joining threads we can safely delete the threads, the answer we want is stored inside sums.
+	threads.clear();
+
+	return result;
+
+}
+
+Matrix column_wise_parallel_masked_image_generator(Matrix & input, Matrix & convolution, int number_of_threads)
```

Comment:
Consider again replacing Matrix&'s with const Matrix&, if that argument is not going to be changed.

Reply:
It just adds errors as I don't know what I am doing, I am just adding const & , but turns out it isn't as easy as this